llpc/lower/llpcSpirvLowerRayTracingIntrinsics.cpp

```diff
@@ -183,4 +211,108 @@ void SpirvLowerRayTracingIntrinsics::createConvertF32toF16(Function *func, unsig
+
+  assert(func->getBasicBlockList().size() == 1);
+  (*func->begin()).eraseFromParent();
+
+  BasicBlock *entryBlock = BasicBlock::Create(m_builder->getContext(), "", func);
+  m_builder->SetInsertPoint(entryBlock);
+  auto argIt = func->arg_begin();
+
+  Type *gpuVaTy = m_builder->getInt64Ty();
+  Type *valueTy = is64Ty ? cast<Type>(FixedVectorType::get(m_builder->getFloatTy(), 2)) : m_builder->getFloatTy();
+  // TODO: Remove this when LLPC will switch fully to opaque pointers.
+  assert(IS_OPAQUE_OR_POINTEE_TYPE_MATCHES(argIt->getType(), gpuVaTy));
+  Value *gpuVa = m_builder->CreateLoad(gpuVaTy, argIt++);
+  Value *offset = m_builder->CreateLoad(m_builder->getInt32Ty(), argIt++);
+  // TODO: Remove this when LLPC will switch fully to opaque pointers.
+  assert(IS_OPAQUE_OR_POINTEE_TYPE_MATCHES(argIt->getType(), valueTy));
+  Value *value = m_builder->CreateLoad(valueTy, argIt);
+
+  if (is64Ty)
+    value = m_builder->CreateBitCast(value, m_builder->getDoubleTy());
+
+  Type *gpuVaAsPtrTy = Type::getInt8PtrTy(m_builder->getContext(), SPIRAS_Uniform);
+  auto gpuVaAsPtr = m_builder->CreateIntToPtr(gpuVa, gpuVaAsPtrTy);
+  // Create GEP to get the byte address with byte offset
+  gpuVaAsPtr = m_builder->CreateGEP(m_builder->getInt8Ty(), gpuVaAsPtr, offset);
+  Type *gpuVaPtrTy = is64Ty ? Type::getDoublePtrTy(m_builder->getContext(), SPIRAS_Uniform)
+                            : Type::getFloatPtrTy(m_builder->getContext(), SPIRAS_Uniform);
+  gpuVaAsPtr = m_builder->CreateBitCast(gpuVaAsPtr, gpuVaPtrTy);
+
+  AtomicRMWInst::BinOp binOp = isMin ? AtomicRMWInst::FMin : AtomicRMWInst::FMax;
+
+  Value *ret =
+      m_builder->CreateAtomicRMW(binOp, gpuVaAsPtr, value, MaybeAlign(), AtomicOrdering::Monotonic, SyncScope::System);
+  m_builder->CreateRet(ret);
+}
+
+// =====================================================================================================================
+// Create AmdExtAtomicLdsFMin, AmdExtAtomicLdsFMax
+//
+// @param func : Function to create
+// @param isMin : Whether is Min operation, otherwise Max operation
+void SpirvLowerRayTracingIntrinsics::createAtomicLdsFMinMax(Function *func, bool isMin) {
+  // float AmdExtAtomicLdsFMin(uint lds[], uint offset, float value)
+  assert(func->getBasicBlockList().size() == 1);
+  (*func->begin()).eraseFromParent();
+
+  BasicBlock *entryBlock = BasicBlock::Create(m_builder->getContext(), "", func);
+  m_builder->SetInsertPoint(entryBlock);
+  auto argIt = func->arg_begin();
+
+  // NOTE: HLSL does not allow function parameter to be groupshared (StorageClassWorkgroup), but can call a function
+  // with a groupshared argument. For this intrinsic, we can always assume the parameter is groupshared so mutate its
+  // type here to resolve the mismatch.
+  // Known issue: SPIRVReader will assert because of the described mismatch function call.
+  argIt->mutateType(PointerType::getWithSamePointeeType(dyn_cast<PointerType>(argIt->getType()), SPIRAS_Local));
```

Comment:
Yeah, please fix that known issue :)

Reply:
Not sure how to fix this :(

We basically do:
groupshared float lds[];
call AmdExtAtomicLdsFMin(lds, index, value);

while the function definition looks like:
float AmdExtAtomicLdsFMin(out lds[LDS_SIZE] _(without groupshared)_, uint index, float value)

And that mismatch stays in SPIR-V, where the argument of OpFunctionCall and parameter of OpFunction has different storage class. Assertion will be triggered when [creating a CallInst](https://github.com/GPUOpen-Drivers/llpc/blob/070a700867b8119ad531e0935906e3785a382c03/llpc/translator/lib/SPIRV/SPIRVReader.cpp#L5012).

Reply:
> Does this mean that the input SPIR-V fails validation?

Yes.

> Would it be possible to fix that in DXC or with some preprocessing?

I don't think so, as HLSL does not allow `groupshared` on a function parameter at the first place.